Sources/swift-format/Utilities/DiagnosticPrinting.swift

```diff
@@ -0,0 +1,37 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Dispatch
+import Foundation
+import TSCBasic
+
+/// The queue used to synchronize printing uninterrupted diagnostic messages.
+private let printQueue = DispatchQueue(label: "com.apple.swift-format.printDiagnosticToStderr")
+
+/// Prints a diagnostic to standard error.
+func printDiagnosticToStderr(_ diagnostic: TSCBasic.Diagnostic) {
+  printQueue.sync {
```

Comment:
I wonder if there would be any appreciable performance difference using `async` here, especially in parallel mode. In the end, we would need to flush the print queue though (so all diagnostics get printed before exiting), which I guess would have to be handled in the Linter and Format frontends. Might not be worth the added complexity.

Reply:
It's certainly worth trying out as follow-up work to see if there's a meaningful difference.